File: app/src/main/java/com/halfplatepoha/jisho/v2/search/SearchFragment.java
package com.halfplatepoha.jisho.v2.search;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import com.halfplatepoha.jisho.R;
import com.halfplatepoha.jisho.analytics.Analytics;
import com.halfplatepoha.jisho.base.BaseFragment;
import com.halfplatepoha.jisho.utils.IConstants;
import com.halfplatepoha.jisho.v2.detail.DetailsActivity;
import com.halfplatepoha.jisho.view.AVLoadingIndicatorView;
import com.halfplatepoha.jisho.view.CustomEditText;

import java.util.Timer;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;

/**
 * Created by surjo on 20/12/17.
 */

public class SearchFragment extends BaseFragment<SearchContract.Presenter> implements SearchContract.View, TextView.OnEditorActionListener, SwitchToOfflineDialog.Listener, TextWatcher {

    public static final String EXTRA_SEARCH_STRING = "extra_search_string";
    public static final String EXTRA_SOURCE = "extra_source";

    @BindView(R.id.etSearch)
    CustomEditText etSearch;

    @BindView(R.id.rlWords)
    RecyclerView rlWords;

    @BindView(R.id.btnClear)
    ImageButton btnClear;

    @BindView(R.id.ivSearch)
    ImageView ivSearch;

    @BindView(R.id.zero_offline_search)
    View zeroOfflineSearch;

    @BindView(R.id.loader)
    AVLoadingIndicatorView loader;

    @Inject
    EntriesAdapter searchAdapter;

    private Timer timer;

    @Override
    protected int getLayoutRes() {
        return R.layout.fragment_search;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

//        loader.setRepeatMode(LottieDrawable.INFINITE);
//        loader.setAnimation("loader.json");

        etSearch.setOnEditorActionListener(this);
        etSearch.addTextChangedListener(this);

        rlWords.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        rlWords.setAdapter(searchAdapter);
    }

    @Override
    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
        if (actionId == EditorInfo.IME_ACTION_SEARCH) {
            Analytics.getInstance().recordSearch(etSearch.getText().toString());

            presenter.searchOnEditorAction(v.getText().toString());
            hideKeyboard();
            return true;
        }
        return false;
    }

    @Override
    public void openDetails(String japanese, String furigana) {
        startActivity(DetailsActivity.getLaunchIntent(getContext(), japanese, furigana));
    }

    @Override
    public void showSpinner() {
        loader.show();
    }

    @Override
    public void hideSpinner() {
        loader.hide();
    }

    @Override
    public void changeSearchListOrientation(int currentOrientation) {
        if(currentOrientation == EntriesAdapterPresenter.TYPE_VERTICAL)
            rlWords.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        else
            rlWords.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
    }

    @OnClick(R.id.btnReportOffline)
    public void clickOffline() {
        presenter.report(etSearch.text());
    }

    @Override
    public void showVerticalSpace() {
//        verticalSpace.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideVerticalSpace() {
//        verticalSpace.setVisibility(View.GONE);
    }

    @Override
    public void openGmailForError(String title) {
        Intent intent = new Intent(Intent.ACTION_SENDTO)
                .setData(Uri.parse("mailto:" + IConstants.DEVELOPER_EMAIL))
                .putExtra(Intent.EXTRA_SUBJECT, title);
        startActivity(intent);
    }

    @Override
    public void hideZeroOffline() {
        zeroOfflineSearch.setVisibility(View.GONE);
    }

    @Override
    public void showZeroOffline() {
        zeroOfflineSearch.setVisibility(View.VISIBLE);
    }

    @Override
    public void showOfflineSwitchConfirmation() {
        SwitchToOfflineDialog dlg = new SwitchToOfflineDialog();
        dlg.setListener(this);
        dlg.show(mChildFragmentManager, SwitchToOfflineDialog.TAG);
    }

    @Override
    public void onConfirm() {
        presenter.onOfflineSwitchConfirm(etSearch.text());
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        presenter.searchOnTextChange(s.toString(), before, count);
    }

    @Override
    public void afterTextChanged(Editable s) {}
}

File: app/src/main/java/com/halfplatepoha/jisho/v2/search/SearchFragment.kt
package com.halfplatepoha.jisho.v2.search

import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.Toast
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.halfplatepoha.jisho.R
import com.halfplatepoha.jisho.v2.base.BaseFragment
import com.halfplatepoha.jisho.v2.detail.DetailActivity
import com.halfplatepoha.jisho.v2.viewmodel.SearchViewModel
import kotlinx.android.synthetic.main.fragment_search.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.koin.android.scope.currentScope
import org.koin.android.viewmodel.ext.android.viewModel
import kotlin.coroutines.CoroutineContext

class SearchFragment : BaseFragment(), CoroutineScope {

    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main

    val searchFragmentViewModel: SearchFragmentViewModel by currentScope.viewModel(this)

    val searchViewModel: SearchViewModel by viewModel()

    val searchAdapter: SearchAdapter by currentScope.inject()

    override val layoutId: Int
        get() = R.layout.fragment_search

    override fun initUi() {
        searchAdapter.isVerticalScroll = true

        rlWords.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        rlWords.adapter = searchAdapter
    }

    override fun setListeners() {
        searchAdapter.listener = {searchFragmentViewModel.onSearchResultSelected(it)}

        btnClear.setOnClickListener { searchFragmentViewModel.clearClicked() }

        etSearch.addTextChangedListener(object : TextWatcher {

            private var previousSearch: String? = ""

            override fun afterTextChanged(s: Editable?) = Unit

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = Unit

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                s?.run {
                    /*
                        do dynamic search only when length is more than 2.
                        else we'll let user click the search icon explicitly
                     */
                    if(length >= 2) {
                        val currentSearch = toString()
                        if (currentSearch.equals(previousSearch)) {
                            return
                        }

                        previousSearch = currentSearch

                        launch {
                            /*
                                debounce timeout
                             */
                            delay(300)
                            if (!currentSearch.equals(previousSearch))
                                return@launch

                            searchFragmentViewModel.onSearched(currentSearch)
                            searchViewModel.onSearched(currentSearch)
                        }
                    }
                }
            }
        })

    }

    override fun setViewModelObservers() {

        searchFragmentViewModel.searchText.observe(this, Observer { etSearch.setText(it) })

        searchFragmentViewModel.selectedSearchResult.observe(this, Observer {
            startActivity(DetailActivity.getLaunchIntent(context!!, it))
        })

        searchFragmentViewModel.searchResults.observe(this, Observer { searchAdapter.searchResults = it })

        searchViewModel.searchResult.observe(this, Observer { searchFragmentViewModel.onSearchResultsReceived(it) })

        searchFragmentViewModel.crossIconVisibility.observe(this, Observer { btnClear.visibility = if(it) View.VISIBLE else View.GONE })
    }

}